### PR TITLE
Horizontal legend traces wrap to new lines

### LINKED ISSUE #769: Horizontal legend items not visible when many legend items
Was pointed out in #766 that scrolling not intended for `layout.legend.orientation = 'h'`.  Example workaround is to use [legend groups](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/legend_horizontal_groups.json). 

This is not a great solution, however, as it means clicking the legend group hides/shows ALL traces in the group (see [jsfiddle](https://jsfiddle.net/psalmody/c9x3xzg9/)).

Proposing that plotly.js implement something similar to ggplot2 which automatically wraps additional legend items on horizontal legend orientation. Especially on mobile portrait orientation, this allows the graph data to take up the entire width of the screen and push the legend below the graph which is much preferable in my view.
kpi-2-raceretention

### PR DESCRIPTION
**Pull request by psalmody**
Resolves #769 

Iterate through traces, get maximum width, use margin and trace width to calculate if a new row is needed. Thanks @n-riesco for reviewing.

new horizontal legend images

### CODE REVIEW

**Comment by etpinard** [comment]
@psalmody looking good. Thanks a lot for doing this.

To complete this PR:

- Can you add an image test mock in the [`test/image/mocks/`](https://github.com/plotly/plotly.js/tree/master/test/image/mocks) directory showing off this new functionality? If you're feeling adventurous, you can try to generate the baseline yourself in our image test docker container (docs [here](https://github.com/plotly/plotly.js/blob/master/test/image/README.md)).
- Can you add a jasmine test case in [`test/jasmine/tests/legend_test.js`](https://github.com/plotly/plotly.js/blob/11dba5ff4684401c722bdb82e72be19326481c2b/test/jasmine/tests/legend_test.js#L507) making sure that changing the `legend.orientation` of a legend with many items makes it properly toggle between the scrollable and multi-column view

Let me know if you have any questions.

**Comment by psalmody** [comment]
@etpinard I added the mock and test. On the mock, I added more traces in the legend for better visibility (particularly in vertical legend orientation). I setup docker (I'm on Win10 and new to docker) and had some problems. Even extracting the linux commands from `tasks/test_image.sh` I'm having issues - appears that `/var/www/streambed/image_server/plotly.js` is empty in the docker container.

**Comment by etpinard** [comment]
@psalmody 

Here's the [link](https://2061-45646037-gh.circle-artifacts.com/0//home/ubuntu/plotly.js/build/test_images/legend_horizontal_autowrap.png) image baseline associated with the mock you added. Simpy download it and move it to `test/image/baselines/legend_horizontal_autowrap.png`.

> Also - the readme makes it seem like this works with windows, but the tasks/test_image.sh file is not written for windows cmd.

Yeah that's correct. I've been wanting to rewrite the shell scripts found in `tasks/` in node as opposed ever since the FOSS release. My apologies. I've opened up an issue (https://github.com/plotly/plotly.js/issues/794) this should be completed soon.

File: test/jasmine/tests/legend_test.js [hunk 1]
```diff
@@ -540,3 +540,73 @@ describe('legend relayout update', function() {
...
+    it('changing to vertical orientation', function(done) {
+        //change mock layout and refresh graph
+        mock.layout.legend.orientation = 'v';
+        mock.layout.legend.y = 1;
+        done();
+    });
+
+    it('changed to vertical, scrollbar appeared', function(done) {
+        //changed orientation to vertical
+        expect(gd._fullLayout.legend.orientation).toBe('v');
```
Comment (etpinard) [inline test/jasmine/tests/legend_test.js]: @psalmody can you rewrite this test as 

```js
Plotly.relayout(gd, 'legend.orientation', 'v').then(function() {
  /* assertions */

 return Plotly.relayout(gd, 'legend.orientation', 'h');
}.then(function() {
  /* assertions (back to base case) /*
});
```

and remove the assertion-less `it('changing to vertical orientation', ...` above.

**Comment by etpinard** [comment]
@psalmody Great. 

I'll merge this PR in as soon as https://github.com/plotly/plotly.js/pull/786#discussion_r72816371 is addressed. Thanks again for this PR!

**Comment by psalmody** [comment]
@etpinard I updated the test (I think) the way you would like it. However, I had to completely destroy the graph and recreate in order to get the legend scrollbar to appear. Using `Plotly.relayout` does change the orientation but it puts the vertical legend overlapping the plot and doesn't show the scrollbar until you use the mouse wheel inside the area (at least on chrome). I'm guessing there are additional layout specifications that must be added by orientation vertical that would also need to be updated in `Plotly.relayout`.

**Comment by etpinard** [comment]
> .. recreate in order to get the legend scrollbar to appear


Oh. Looks like you discovered some buggy `relayout` behaviour in scrollable legends. Thanks for reporting :beers: . I made an issue about it here: https://github.com/plotly/plotly.js/issues/808

>  Using Plotly.relayout does change the orientation but it puts the vertical legend overlapping the plot and doesn't show the scrollbar until you use the mouse wheel inside the area (at least on chrome).

`Plotly.relayout` should work here. But from what I understand, a lot of legend relayout cases (that have nothing to do with your PR) still need to be worked out.

So, I suggest removing the jasmine test you added in https://github.com/plotly/plotly.js/pull/786/commits/3a2cf7ba56337542fe17467ca4c934b9fb374e5e and patched in https://github.com/plotly/plotly.js/pull/786/commits/c31eb68cc0f9cfe0e7ac147b00de7a4e1abc79a1 and instead add a jasmine test case testing some that `relayout` with some style attribute e.g. `legend.bgcolor` works as expected for mulit-line horizontal legends. Your patch in https://github.com/plotly/plotly.js/pull/786/commits/c31eb68cc0f9cfe0e7ac147b00de7a4e1abc79a1 by calling `Plotly.plot` in sequence essentially makes the jasmine test the same code paths as the test image test.

**Comment by psalmody** [comment]
> Your patch in c31eb68 by calling Plotly.plot in sequence essentially makes the jasmine test the same code paths as the test image test.

Ah - that makes sense. Removed old tests and created new test as recommended.

**Comment by etpinard** [comment]
@psalmody Great. Thanks very much for this PR. You get my vote for PR of the week :trophy:

**Comment by Braintelligence** [comment]
@psalmody Thank you for this PR! Do you have any idea why https://github.com/plotly/plotly.js/issues/3175 is happening, though? =/ I can't pinpoint where the code is at fault by my own 😿.

**Pull request merged**